File: 配套资源/配套代码/程序设计竞赛专题挑战教程-代码/第4章双向排序.cpp
// 双向排序 2021年第十二届省赛C/C++大学B组、研究生组，Java大学A组、B组、C组、研究生组
//lanqiaoOJ 1458 

#include <bits/stdc++.h>
using namespace std;
#define x first
#define y second
const int N = 100010;
int ans[N];
pair<int, int> stk[N];
int main() {
    int n, m;  cin>>n>>m;
    int top = 0;    // 栈顶
    while (m--){    //读操作，并且合并
        int p,q; scanf("%d%d", &p, &q);        //这里不用cin，慢
        if (p==0)  {                           //操作0：降序
            while(top && stk[top].x == 0)      //合并连续的0操作，取最大
                q = max(q, stk[top--].y);
            while(top >= 2 && stk[top-1].y <= q) //0、1交替
            //如果当前的操作0比上一次的操作0范围大，将上一次操作0和操作1删除
                top -= 2;
            stk[++top] = {0, q};   //存本次最佳操作
        }
        else if (top) {  //操作1 &&且操作0已经进行过（操作1第一个用没效果）
            while (top && stk[top].x == 1)
                q = min(q, stk[top--].y);
            while (top >= 2 && stk[top-1].y >= q)
                top -= 2;
            stk[++top] = {1, q};
        }
    }
    int k=n, L=1, R=n;        //用k处理固定的数字
    for (int i=1; i<=top; i++ ) {
        if (stk[i].x == 0)    //0降序操作，固定右边的数字
            while (R>stk[i].y && L<=R) ans[R--] = k--;
        else                  //1升序操作，固定左边的数字
            while (L<stk[i].y && L<=R) ans[L++] = k--;
        if (L>R) break;
    }
    if (top % 2)             //最后一次操作是0操作，补上中间的数字
        while(L<=R) ans[L++] = k--;
    else                     //最后一次操作是1操作
        while(L<=R) ans[R--] = k--;
    for (int i=1; i<=n; i++)  printf("%d ", ans[i]);
    return 0;
}

File: 配套资源/配套代码/程序设计竞赛专题挑战教程-代码/第4章双向排序.py
#双向排序 2021年第十二届省赛C/C++大学B组、研究生组，Java大学A组、B组、C组、研究生组
#lanqiaoOJ 1458
   
stk = []      #空栈
n,m = map(int,input().split())
for i in range(m):
    p,q = map(int,input().split())
    if p == 0:                        #0操作
       while stk and stk[len(stk)-1][0]==0:
           q = max(q,stk.pop()[1])
       while len(stk)>=2 and stk[len(stk)-2][1] <= q:
           stk.pop()
           stk.pop()
       stk.append([0,q])
    elif stk:
       while stk and stk[len(stk)-1][0]==1:
           q = min(q,stk.pop()[1])
       while len(stk)>=2 and stk[len(stk)-2][1] >= q:
            stk.pop()
            stk.pop()
       stk.append([1,q])
k,L,R = n,1,n
ans = [0]*(n+1)
for x,y in stk:   #逐个处理栈元素    
    if x == 0:
        while R>y and L<=R:  ans[R]=k; k=k-1; R=R-1
    else:
        while L<y and L<=R:  ans[L]=k; k=k-1; L=L+1
    if L>R: break
if x == 0:
        while L<=R: ans[L]=k; k=k-1; L=L+1
else :
        while L<=R: ans[R]=k; k=k-1; R=R-1
for i in range(1,n+1):  print(ans[i],end=' ')
